File: src/main/java/com/wesker926/leetcode/algorithms/p0630/s1/solution.go
// Package s1
// @author: wesker.gh
// @date: 2021/12/14
// @description: 贪心 + 优先级队列
package s1

import (
	"container/heap"
	"runtime/debug"
	"sort"
)

func init() {
	debug.SetGCPercent(-1)
}

func scheduleCourse(courses [][]int) int {
	sort.Slice(courses, func(i, j int) bool {
		return courses[i][1] < courses[j][1]
	})
	h, total := &Heap{}, 0
	for _, course := range courses {
		if t := course[0]; total+t <= course[1] {
			total += t
			heap.Push(h, t)
		} else if h.Len() > 0 && t < h.IntSlice[0] {
			total += t - h.IntSlice[0]
			h.IntSlice[0] = t
			heap.Fix(h, 0)
		}
	}
	return h.Len()
}

type Heap struct {
	sort.IntSlice
}

func (h Heap) Less(i, j int) bool {
	return h.IntSlice[i] > h.IntSlice[j]
}

func (h *Heap) Push(v interface{}) {
	h.IntSlice = append(h.IntSlice, v.(int))
}

func (h *Heap) Pop() interface{} {
	a := h.IntSlice
	v := a[len(a)-1]
	h.IntSlice = a[:len(a)-1]
	return v
}

File: src/main/java/com/wesker926/leetcode/algorithms/p0630/s1/Solution.java
package com.wesker926.leetcode.algorithms.p0630.s1;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

/**
 * @author wesker.gh
 * @date 2021/12/14
 * @description 贪心 + 优先级队列
 */
public class Solution {
    public int scheduleCourse(int[][] courses) {
        Arrays.sort(courses, Comparator.comparingInt(o -> o[1]));

        int total = 0;
        PriorityQueue<Integer> queue = new PriorityQueue<>((o1, o2) -> o2 - o1);
        for (int[] course : courses) {
            if (course[0] + total <= course[1]) {
                total += course[0];
                queue.offer(course[0]);
            } else if (!queue.isEmpty() && queue.peek() > course[0]) {
                total += course[0] - queue.poll();
                queue.offer(course[0]);
            }
        }
        return queue.size();
    }
}
